internal/pkg/print/debug.go

```diff
@@ -57,6 +81,147 @@ func BuildDebugStrFromSlice(inputSlice []string) string {
 	return fmt.Sprintf("[%s]", sliceStr)
 }
 
+// buildHeaderMap converts a map to a user-friendly string representation.
+// This function also filters the headers based on the includeHeaders parameter.
+// If includeHeaders is empty, the default header filters are used.
+func buildHeaderMap(headers http.Header, includeHeaders []string) map[string]any {
+	headersMap := make(map[string]any)
+	for key, values := range headers {
+		headersMap[key] = strings.Join(values, ", ")
+	}
+
+	var headersToInclude []string
+
+	if len(includeHeaders) == 0 {
+		headersToInclude = defaultDebugIncludeHeaders
+	} else {
+		headersToInclude = includeHeaders
+	}
+
+	for key := range headersMap {
+		if slices.Contains(headersToInclude, key) {
+			continue
+		}
+		delete(headersMap, key)
+	}
+
+	return headersMap
+}
+
+// BuildDebugStrFromHTTPRequest converts an HTTP request to a user-friendly string representation.
+// This function also receives a list of headers to include in the output, if empty, the default headers are used.
+// The return value is a list of strings that should be printed separately.
+func BuildDebugStrFromHTTPRequest(req *http.Request, includeHeaders []string) ([]string, error) {
+	if req == nil {
+		return nil, fmt.Errorf("request is nil")
+	}
+	if req.URL == nil || req.Proto == "" || req.Method == "" {
+		return nil, fmt.Errorf("request is invalid")
+	}
+
+	status := fmt.Sprintf("request to %s: %s %s", req.URL, req.Method, req.Proto)
+
+	headersMap := buildHeaderMap(req.Header, includeHeaders)
+	headers := fmt.Sprintf("request headers: %v", BuildDebugStrFromMap(headersMap))
+
+	if req.Body == nil {
+		return []string{status, headers}, nil
+	}
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return []string{status, headers}, fmt.Errorf("read request body: %w", err)
+	}
+	var bodyMap map[string]any
+	if len(body) != 0 {
+		if err := json.Unmarshal(body, &bodyMap); err != nil {
+			return []string{status, headers}, fmt.Errorf("unmarshal request body: %w", err)
+		}
+	}
+
+	// restore body
+	req.Body = io.NopCloser(strings.NewReader(string(body)))
```

Comment:
is there no way of reading the body without emptying it? This seems dangerous.

Maybe you can make a copy of the request (or the body) and use that one

Reply:
I don't think there is a way around it. I have refactored the implementation slightly to match the go std library's implementation, as described in the comments above the new functions.

Reply:
as discussed offline, let's use the `drainBody` function